pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/GeometrySerializer.java

```diff
@@ -0,0 +1,528 @@
+    return fromID(typeByte & GeometryUtils.GEOGRAPHY_GET_MASK);
+  }
+
+  private static Geometry readPolyline(ByteBuffer byteBuffer, boolean multitype, GeometryFactory factory) {
+    int partCount = byteBuffer.getInt();
+    if (partCount == 0) {
+      if (multitype) {
+        return factory.createMultiLineString();
+      }
+      return factory.createLineString();
+    }
+
+    int pointCount = byteBuffer.getInt();
+    int[] startIndexes = new int[partCount];
+    for (int i = 0; i < partCount; i++) {
+      startIndexes[i] = byteBuffer.getInt();
+    }
+
+    int[] partLengths = new int[partCount];
+    if (partCount > 1) {
+      partLengths[0] = startIndexes[1];
+      for (int i = 1; i < partCount - 1; i++) {
+        partLengths[i] = startIndexes[i + 1] - startIndexes[i];
+      }
+    }
+    partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
+
+    LineString[] lineStrings = new LineString[partCount];
+
+    for (int i = 0; i < partCount; i++) {
+      lineStrings[i] = factory.createLineString(readCoordinates(byteBuffer, partLengths[i]));
+    }
+
+    if (multitype) {
+      return factory.createMultiLineString(lineStrings);
+    }
+    Preconditions.checkArgument(lineStrings.length == 1, "The remaining line string must have only one node");
+    return lineStrings[0];
+  }
+
+  private static Geometry readPolygon(ByteBuffer byteBuffer, boolean multitype, GeometryFactory factory) {
+    int partCount = byteBuffer.getInt();
+    if (partCount == 0) {
+      if (multitype) {
+        return factory.createMultiPolygon();
+      }
+      return factory.createPolygon();
+    }
+
+    int pointCount = byteBuffer.getInt();
+    int[] startIndexes = new int[partCount];
+    for (int i = 0; i < partCount; i++) {
+      startIndexes[i] = byteBuffer.getInt();
+    }
+
+    int[] partLengths = new int[partCount];
+    if (partCount > 1) {
+      partLengths[0] = startIndexes[1];
+      for (int i = 1; i < partCount - 1; i++) {
+        partLengths[i] = startIndexes[i + 1] - startIndexes[i];
+      }
+    }
+    partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
+
+    LinearRing shell = null;
+    List<LinearRing> holes = new ArrayList<>();
+    List<Polygon> polygons = new ArrayList<>();
+    try {
+      for (int i = 0; i < partCount; i++) {
+        Coordinate[] coordinates = readCoordinates(byteBuffer, partLengths[i]);
+        if (isClockwise(coordinates)) {
+          // next polygon has started
+          if (shell != null) {
+            polygons.add(factory.createPolygon(shell, holes.toArray(new LinearRing[0])));
+            holes.clear();
+          }
+          shell = factory.createLinearRing(coordinates);
+        } else {
+          holes.add(factory.createLinearRing(coordinates));
+        }
+      }
+      polygons.add(factory.createPolygon(shell, holes.toArray(new LinearRing[0])));
+    } catch (IllegalArgumentException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)